Implement adaptive FCI chunks padding and create a new FCIChunksYAMLReader class

ameraner opened the pull request:
This PR implements a new YAMLReader to handle FCI chunks. This was needed since a new functionality needed to be implemented: the chunk sizes of the FCI files is configurable by the data producer, hence the gaps to be padded can be of variable sizes. The new YAMLReader computes the chunks to be padded in an adaptive way. 
To keep the GEOSegmentYAMLReader code clean and generic, it was decided to create this new class. 

New tests have been added. The old tests were passing after minor adaptations for the new code structure (e.g. change from function to class method).

Code check with real data:
```python
from satpy.scene import Scene
from glob import glob
from matplotlib import pyplot as plt

path_to_testdata = glob("/path_to_test_data_with_missing_chunks" + "/*BODY*.nc")
scn = Scene(filenames=path_to_testdata, reader=['fci_l1c_nc'])

scn.load(['vis_04'], calibration='radiance', upper_right_corner='NE')

v = scn['vis_04'].values
plt.figure()
plt.imshow(v)
plt.colorbar()
plt.show(block=False)

scn_r = scn.resample('msg_seviri_fes_3km')

v = scn_r['vis_04'].values
plt.figure()
plt.imshow(v)
plt.colorbar()
plt.show(block=False)
```
![image](https://user-images.githubusercontent.com/49722768/144617517-4b390363-f4d2-41bc-a264-e237136af106.png)

Both images look the same, showing that the full-disk padding works and also the geolocation of the padded chunks is correct.


Similarly for SEVIRI HRIT:
```python
path_to_testdata = '/path_to_test_data_with_missing_chunks/' 
scn = Scene(filenames=glob(path_to_testdata + "H*"), reader=['seviri_l1b_hrit'])

scn.load([channel_name], upper_right_corner='NE')

plt.figure()
plt.imshow(scn[channel_name].values)
plt.show(block=False)

scn_r = scn.resample('msg_seviri_fes_1km')

plt.figure()
plt.imshow(scn_r[channel_name].values)
plt.colorbar()
plt.show(block=False)
```
![image](https://user-images.githubusercontent.com/49722768/144619055-f2859ffb-89be-4940-a172-06e79ca39f31.png)

 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [x] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

PS: On a side, it adds a # type: ignore for the cached_property imports, to avoid issues with the `mypy` git hooks.

gerritholl commented:
> the chunk sizes of the FCI files is configurable by the data producer, hence the gaps to be padded can be of variable sizes

I appreciate it's motivated by FCI, but should the reader perhaps have a more generic name, `VariableChunksYAMLReader` or so?  Then write in the documentation that it's motivated by FCI.

ameraner commented:
Thank you @gerritholl for your comment, indeed the new YAMLReader can also be formulated generically, that's a good idea to keep the YAMLReaders as instrument-independent as possible. 
I will reformulate the function names/descriptions, and put the extraction of the chunk position information inside the FCI reader in a separate, standardized method, so that another reader can implement the same method and reuse the YAMLReader.

pnuu commented on satpy/readers/yaml_reader.py:
```diff
@@ -1221,25 +1207,94 @@ def _load_dataset(dsid, ds_info, file_handlers, dim='y', pad_data=True):
+        area_defs = {}
 
-    return area_def
+        for segment in range(available_segments[0], expected_segments + 1):
+            try:
+                idx = available_segments.index(segment)
+                fh = file_handlers[idx]
+                area = fh.get_area_def(dsid)
+            except ValueError:
+                logger.debug("Padding to full disk with segment nr. %d", segment)
+
+                new_height_proj_coord, new_height_px = self._get_new_areadef_heights(area, seg_size,
+                                                                                     segment_n=segment,
+                                                                                     filetype=filetype)
+                new_ll_y = area.area_extent[1] + new_height_proj_coord
+                new_ur_y = area.area_extent[1]
+                fill_extent = (area.area_extent[0], new_ll_y,
+                               area.area_extent[2], new_ur_y)
+                area = AreaDefinition('fill', 'fill', 'fill', area.crs,
+                                      seg_size[1], new_height_px,
+                                      fill_extent)
+
+            area_defs[segment] = area
+            seg_size = area.shape
```
Coudl this be a separate function? And further have the exception handling in yet another function.

ameraner replied:
I refactored the exception handling, so that the method is much slimmer now. But I kept the try/except here, as refactoring further would introduce methods with many arguments, so not sure if it helps readability/makes much sense in the new code. What do you think?

pnuu replied:
The whole `for` loop could be a separate method, say `return self._get_area_def_segments(filehandler, dsid, available_segments, expected_segments)` and move the `segment = None` initialisation also inside it.

pnuu commented on satpy/readers/yaml_reader.py:
```diff
@@ -1221,25 +1207,94 @@ def _load_dataset(dsid, ds_info, file_handlers, dim='y', pad_data=True):
+            logger.debug("Padding segment %d to full disk.",
+                         segment)
+
+            new_height_proj_coord, new_height_px = self._get_new_areadef_heights(area, seg_size,
+                                                                                 segment_n=segment,
+                                                                                 filetype=filetype)
+            new_ll_y = area.area_extent[3]
+            new_ur_y = area.area_extent[3] - new_height_proj_coord
+            fill_extent = (area.area_extent[0], new_ll_y,
+                           area.area_extent[2], new_ur_y)
+            area = AreaDefinition('fill', 'fill', 'fill',
+                                  area.crs,
+                                  seg_size[1], new_height_px,
+                                  fill_extent)
+            area_defs[segment] = area
+            seg_size = area.shape
```
This structure is very similar to the one in the exception block in `_pad_later_segments_area`, maybe these could be combined to one function?

ameraner replied:
Done

pnuu commented on satpy/readers/yaml_reader.py:
```diff
@@ -1374,22 +1354,141 @@ def _get_empty_segment_with_height(empty_segment, new_height, dim):
+    """GEOVariableSegmentYAMLReader for handling chunked/segmented GEO products with segments of variable height.
 
-    return chunk_height
+    This YAMLReader ovverrides parts of the GEOSegmentYAMLReader to account for formats where the segments can
```
```suggestion
    This YAMLReader overrides parts of the GEOSegmentYAMLReader to account for formats where the segments can
```

ameraner replied:
Fixed, thanks.

pnuu commented on satpy/readers/yaml_reader.py:
```diff
@@ -1374,22 +1354,141 @@ def _get_empty_segment_with_height(empty_segment, new_height, dim):
+    def _extract_segment_location_dicts(self, created_fhs):
+        self.segment_infos = dict()
+        for filetype, filetype_fhs in created_fhs.items():
+            # initialise the segment info for this filetype
+            exp_segment_nr = filetype_fhs[0].filetype_info['expected_segments']
+            width_to_grid_type = _get_width_to_grid_type(filetype_fhs[0].get_segment_position_info())
+            self.segment_infos.update({filetype: {'available_segment_infos': [],
+                                                  'expected_segments': exp_segment_nr,
+                                                  'width_to_grid_type': width_to_grid_type}})
```
This could be a separate method, for example `self._extract_initial_segment_position_info(filetype, filetype_fhs)`.

pnuu commented on satpy/readers/yaml_reader.py:
```diff
@@ -1374,22 +1354,141 @@ def _get_empty_segment_with_height(empty_segment, new_height, dim):
+                                                  'expected_segments': exp_segment_nr,
+                                                  'width_to_grid_type': width_to_grid_type}})
+
+            # collect the segment positioning infos for all available segments
+            for fh in filetype_fhs:
+                chk_infos = fh.get_segment_position_info()
+                chk_infos.update({'segment_nr': fh.filename_info['segment'] - 1})
+                self.segment_infos[filetype]['available_segment_infos'].append(chk_infos)
```
This could be a separate method, for example `self._collect_segment_position_infos(filetype, filetype_fhs)`.

ameraner replied:
Done, thanks.

pnuu reviewed (commented):
Nice job! Few comments inline.

mraspaud reviewed (commented):
Thanks for creating this PR, it makes sense to create a new class indeed.

I have a few comments inline, but some more general comments here:
- The amount of code for the geo readers is getting big, should we have a separate modules for these?
- Many places have comments explaining exactly what is happening. I understand and appreciate the intent, however I wonder if in some cases these could be made obsolete by having small function with very descriptive names and/or variables with good names?

ameraner commented on satpy/readers/yaml_reader.py:
```diff
@@ -42,6 +42,7 @@
 from trollsift.parser import globify, parse
 
 from satpy import DatasetDict
+from satpy._compat import cached_property
```
Done, thanks.

ameraner commented on satpy/readers/yaml_reader.py:
```diff
@@ -1221,25 +1207,94 @@ def _load_dataset(dsid, ds_info, file_handlers, dim='y', pad_data=True):
+                               area.area_extent[2], new_ur_y)
+                area = AreaDefinition('fill', 'fill', 'fill', area.crs,
+                                      seg_size[1], new_height_px,
+                                      fill_extent)
```
Done, thanks.

ameraner commented on satpy/readers/yaml_reader.py:
```diff
@@ -1374,22 +1354,141 @@ def _get_empty_segment_with_height(empty_segment, new_height, dim):
+            segment_start_rows[group[n]] = segment_start_rows[group[n - 1]] + proposed_sizes_missing_segments[n] + 1
+        if n != len(group) - 1:
+            segment_end_rows[group[n]] = segment_start_rows[group[n]] + proposed_sizes_missing_segments[n]
+        segment_heights[group[n]] = proposed_sizes_missing_segments[n]
```
Done, thanks, hope it's better now.

ameraner commented on satpy/readers/yaml_reader.py:
```diff
@@ -1374,22 +1354,141 @@ def _get_empty_segment_with_height(empty_segment, new_height, dim):
+        current_fh_segment_nr = chk_info['segment_nr']
+        segment_heights[current_fh_segment_nr] = chk_info[grid_type]['segment_height']
+        segment_start_rows[current_fh_segment_nr] = chk_info[grid_type]['start_position_row']
+        segment_end_rows[current_fh_segment_nr] = chk_info[grid_type]['end_position_row']
```
Done, thanks.

ameraner commented:
@mraspaud @pnuu thank you for your review! I addressed all your points, let me know if you have any more comments. I also added a test to cover an untested overridden method.

mraspaud reviewed (approved):
LGTM

ameraner commented:
Thank you for the review and the merge :)
